src/test/java/spoon/test/trycatch/TryCatchTest.java

```diff
@@ -492,4 +498,80 @@ private <T extends Throwable> CtCatch createCatch(Factory factory, Class<T> type
 			);
 		}
 	}
+
+	@Test
+	void testFieldAsCtResource() {
+		// contract: Since java 9 normal variables, fields, parameters and catch variables are allowed
+		// in try-with-resources
+		CtModel model = createModelFromString("""
+			class ClosableException extends RuntimeException implements AutoCloseable {
+				@Override
+				public void close() throws Exception {}
+			}
+			class Foo {
+				final AutoCloseable field = null;
+				public void bar(AutoCloseable param) {
+					try { }
+					catch (ClosableException e) {
+						AutoCloseable localVar = param;
+						try (e; field; param; localVar; AutoCloseable inside = () -> {}) {
+						} catch (Exception ignored) {}
+					}
+				}
+			}
+			""");
+		CtClass<?> foo = (CtClass<?>) model.getAllTypes()
+			.stream()
+			.filter(it -> it.getSimpleName().equals("Foo"))
+			.findAny()
+			.orElseThrow();
+		CtMethod<?> bar = foo.getMethodsByName("bar").get(0);
+
+		CtCatchVariable<?> e = bar.getElements(new TypeFilter<>(CtCatchVariable.class)).get(0);
+		CtLocalVariable<?> localVar = bar.getElements(new TypeFilter<>(CtLocalVariable.class))
+			.stream()
+			.filter(it -> it.getSimpleName().equals("localVar"))
+			.findAny()
+			.orElseThrow();
+		CtField<?> field = foo.getField("field");
+		CtParameter<?> param = bar.getParameters().get(0);
+		CtLocalVariable<?> inside = bar.getElements(new TypeFilter<>(CtLocalVariable.class))
+			.stream()
+			.filter(it -> it.getSimpleName().equals("inside"))
+			.findAny()
+			.orElseThrow();
+
+		List<CtResource<?>> resources = bar.getElements(new TypeFilter<>(CtTryWithResource.class))
+			.get(0)
+			.getResources();
+		assertThat(resources).containsExactlyInAnyOrder(
+			getRead(field),
+			getRead(param.getReference()),
+			getRead(e.getReference()),
+			getRead(localVar.getReference()),
+			inside
+		);
+	}
+
+	private <T> CtVariableRead<T> getRead(CtVariableReference<T> ref) {
+		return ref.getFactory().<T>createVariableRead().<CtFieldRead<T>>setVariable(ref);
+	}
+
+	private <T> CtVariableRead<T> getRead(CtField<T> field) {
+		Factory factory = field.getFactory();
+		CtFieldRead<T> read = factory.createFieldRead();
+		read.setVariable(field.getReference());
+		read.setTarget(factory.createThisAccess(field.getDeclaringType().getReference()));
+
+		return read;
+	}
+
+	private static CtModel createModelFromString(String code) {
+		Launcher launcher = new Launcher();
+		launcher.getEnvironment().setComplianceLevel(21);
+		launcher.getEnvironment().setShouldCompile(true);
+		launcher.addInputResource(new VirtualFile(code));
+		return launcher.buildModel();
+	}
```

Comment:
One day, we will have a single method for this :(

Reply:
And other lies I tell myself? *despair*